mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -28,94 +24,93 @@ pub(crate) struct PeerManager {
                     timeline_id: 0,
                     is_alive: true,
-                    network_id,
                 })
                 .is_alive = true;
         }
+        is_new_peer
     }
 
-    pub fn disable_peer(&self, peer_id: PeerId) {
+    pub fn disable_peer(&self, peer: PeerNetworkId) {
         if let Some(state) = self
             .peer_info
             .lock()
-            .expect("failed to acquire peer info lock")
-            .get_mut(&peer_id)
+            .expect("failed to get peer info lock")
+            .get_mut(&peer)
         {
             state.is_alive = false;
         }
     }
 
-    // pick_peers
-    // picks peers to broadcast to. Will first try to pick *all* preferred peers (= upstream as defined by config)
-    // if preferred upstream peers < `k = min_broadcast_recipient_count`, pick fallback peers to fill up k
-    pub fn pick_peers(&self) -> PeerInfo {
-        let peer_info = self
-            .peer_info
+    pub fn update_peer_broadcast(&self, peer: PeerNetworkId, timeline_id: u64) {
+        self.peer_info
             .lock()
-            .expect("failed to acquire peer info lock");
-        // pick all preferred peers
-        let mut picked_peers: PeerInfo = peer_info
-            .iter()
-            .filter(|(peer, state)| {
-                state.is_alive
-                    && self
-                        .upstream_config
-                        .is_primary_upstream_peer(state.network_id, **peer)
-            })
-            .map(|(peer, state)| (*peer, state.clone()))
-            .collect();
+            .expect("failed to acquire peer_info lock")
+            .entry(peer)
+            .and_modify(|t| {
+                t.timeline_id = timeline_id;
+            });
+    }
+
+    pub fn is_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_upstream_peer(peer)
+    }
 
-        let picked_peers_count = picked_peers.len();
-        if picked_peers_count < self.min_broadcast_recipient_count {
-            // randomly select fallback peers
-            // TODO add peer scoring schema to use for selecting fallback peers
-            let fallback_peers: PeerInfo = peer_info
-                .iter()
-                .filter(|(peer, state)| !picked_peers.contains_key(&peer) && state.is_alive)
-                .map(|(peer, state)| (*peer, state.clone()))
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    self.min_broadcast_recipient_count - picked_peers_count,
-                )
-                .into_iter()
-                .collect();
+    fn is_primary_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_primary_upstream_peer(peer)
+    }
 
-            picked_peers.extend(fallback_peers);
+    // returns a peer's sync state if the peer is a chosen broadcast peer
+    pub fn get_picked_state(&self, peer: PeerNetworkId) -> Option<PeerSyncState> {
```

Comment:
nit: this function has interesting signature. Can we just check `is_picked_peer` on call site level and here always return PeerSyncState?